### DISCUSSION TITLE
Realtime subscriptions from the server

### DISCUSSION THREAD

**Discussion by TahLoow** [pocketbase/pocketbase#1897] [Q&A] [answered]
Disclosure: I'm not a security expert

My Pocketbase server is in a containerized network with a nextjs server, so DB access is restricted the nextjs server and cannot be accessed from the internet. I want to stream data from the database, to a server side component, and then to a client browser. However, the server side component cannot subscribe to changes because it doesn't have EventSource access. My alternative seems to be interval polling, but subscriptions are just too alluring. I suspect it can be accomplished using the PocketBase framework, but I am unsure. It seems very few people have this question.

My questions are:
Does it make sense for a server to have subscription/streaming? 
Are server-side subscriptions currently achievable (with either the SDK or the PB framework)?

If you have other insights, please let me know.
Many thanks!

**Comment by ganigeorgiev** [comment, marked as answer]
> Does it make sense for a server to have subscription/streaming?

It depends on your use case and the desired workflow. Based on your description it seems that the Next.js server is acting as a proxy/middleware layer for the PocketBase API, and in this case it should be OK to use it this way.

---

> Are server-side subscriptions currently achievable (with either the SDK or the PB framework)?

Yes, in both cases.

If you want to use the SDK and the realtime service in node, you can load an `EventSource` polyfill:

```js
// npm install eventsource --save
import eventsource from 'eventsource';

global.EventSource = eventsource;

...
```
(_this and other node caveats could be found in https://github.com/pocketbase/js-sdk#nodejs-via-npm_)


You can also directly connect to the Realtime APIs following https://pocketbase.io/docs/api-realtime/.
In general, there are 2 actions - the SSE request (_establishes long lived HTTP connection where you'll receive the event messages_) and the subscriptions submission (_regular POST HTTP request for sending/removing the subscribed topics_).

**Reply by TahLoow** [reply]
Got it working, what a pleasant fix! I'll read through those other sources. Thank you for your time. 

Should anyone else come across this:
- I used NextJS 13
- Linter complained `Could not find a declaration file for module 'eventsource'`. Changing to `const eventsource = require("eventsource");` fixed it despite having ES6 modules
